### PR TITLE
#160609555 create, read, update, delete articles

### PR DESCRIPTION
**Pull request by mendozabree**
#### What does this PR do?

Creates articles app where:
- Users can create articles
- Users can view articles
- Users can only edit articles where they are authors
- Users can only delete articles where they are authors

#### Description of Task to be completed?

Currently, users can not create, read, update or delete articles.

This task creates an articles app in the codebase that handles all the above functions.

#### How should this be manually tested?
Check out the background context section below before testing.

- Make migrations: `python manage.py make migrations articles`
- Migrate: `python manage.py migrate`
- Run the server: `python manage.py runserver`
- Open postman and test these endpoints: 
**Note;** some of the endpoints are restricted to authenticated users, hence you must first get the token using a registration or login.

**Create an article (authenticated)**
POST `http://127.0.0.1:8000/api/articles/`
Request body:
```
{
  "article": {
    	"title": "Testing django part 13",
    	"description": "It should work",
    	"body": "I am testing to confirm"
	}
}
```

**Get all articles(not authenticated)**
GET `http://127.0.0.1:8000/api/article/`

**Get specific article (not authenticated)**
GET `http://127.0.0.1:8000/api/articles/testing-django-part-13/`

**Edit article (authenticated)**
PUT `http://127.0.0.1:8000/api/articles/testing-django-part-13/`
Request body:
```
{
  "article": {
    "body": "Confirming put!"
  }
}
```

**Delete article (authenticated)**
DELETE `http://127.0.0.1:8000/api/articles/testing-django-part-13/`

#### What are the relevant pivotal tracker stories?

[#160609555](https://www.pivotaltracker.com/story/show/160609555)

#### Any background context you want to add?
Since we now have profiles for each user, to run this you will have to drop all tables or create a new database to use because this PR works in such a way that a user_id = profile_id.

#### Screenshots
*Response from creating an article*
screen shot 2018-10-09 at 08 07 30

*Response from getting an article*
screen shot 2018-10-09 at 08 14 39

*Response from updating an article*
screen shot 2018-10-09 at 08 16 28

*Response from deleting an article*
screen shot 2018-10-09 at 08 06 35

### CODE REVIEW

**Comment by RonKbS** [comment]
Nice work @mendozabree.
Tried creating the same article twice and to differentiate the two, a 1 was added to the slug of the second. 👍 
I have a question about the /article/ and /articles/ URLs. Would it not be better to have just one URL with the two methods of post and get?

**Comment by mendozabree** [comment]
@RonKbS Thanks.

About the URLs, when I am creating an article I use `generics.CreateAPIView` and for retrieving all articles I use `generics.ListAPIView`. 
The reason I used the two is that Django rest framework provides them to ease work, I think there might be a way to use `APIView` for both but I think it would be tedious and time consuming.Also the two have different `permission_classes`.

**Comment by samuelbwambale** [comment]
@mendozabree The CRUD works fine only that in the PR procedure, when making migrations,  the command should be ``` python manage.py makemigrations articles ``` or else the articles table will not be created

**Comment by MuhanguziDavid** [comment]
Nice work @mendozabree,
However, I could not get a specific article without authentication yet it is specified in the ReadME that no authentication is required for this endpoint.

**Comment by MuhanguziDavid** [comment]
Please use customized error messages with appropriate status codes when a user searches for an article that doesn't exist.

File: authors/apps/articles/admin.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (MuhanguziDavid) [inline authors/apps/articles/admin.py]: Is this file really necessary?

File: authors/apps/articles/renderers.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+class ArticleJSONRenderer(JSONRenderer):
+    """Class for JSON renderer for Artiles app"""
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        """Method to convert data to a specific format
+        :params data
+        :return JSON object with key article"""
+
+        return json.dumps(
```
Comment (MuhanguziDavid) [inline authors/apps/articles/renderers.py]: 14 and 15 can fit on one line.

File: authors/apps/articles/serializers.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+from .models import Article
+
+
+class ArticleSerializer(serializers.ModelSerializer):
+    """Create a new article"""
+
+    class Meta:
+        model = Article
+        fields = ['author', 'title', 'description', 'body',
+                  'createdAt', 'updatedAt', 'slug']
```
Comment (MuhanguziDavid) [inline authors/apps/articles/serializers.py]: Please format this block as follows:
```python
fields = [
    'author', 'title', 'description', 'body',
    'createdAt', 'updatedAt', 'slug'
]
```

File: authors/apps/articles/tests.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        self.username = "johndoe"
+        self.email = "johndoe@gmail.com"
+        self.password = "Password1"
+
+        self.user_data = {"user": {"username": self.username,
+                            "email": self.email,
+                            "password": self.password}}
+
+        self.response = self.client.post(
+            '/api/users/', self.user_data, format="json")
```
Comment (MuhanguziDavid) [inline authors/apps/articles/tests.py]: This block can fit on one line.

File: authors/apps/articles/urls.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+
+from .views import CreateArticle, ArticleRetrieveUpdate, ArticleList
+
+urlpatterns = [
+    path('articles/', CreateArticle.as_view()),
+    path('articles/<str:slug>/', ArticleRetrieveUpdate.as_view()),
+    path('article/', ArticleList.as_view())
+        ]
```
Comment (MuhanguziDavid) [inline authors/apps/articles/urls.py]: Please, push this bracket to the beginning of the line

File: authors/apps/articles/views.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                             "description": serializer.data['description'],
+                             "body": serializer.data['body'],
+                             "createdAt": serializer.data['createdAt'],
+                             "updatedAt": serializer.data['updatedAt'],
+                             "slug": serializer.data['slug']
+                            },
+                 "message": "Success"
+                 },
+                status=status.HTTP_200_OK
+                        )
```
Comment (MuhanguziDavid) [inline authors/apps/articles/views.py]: Please align this bracket with the return keyword.

File: authors/apps/articles/views.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        article = Article.get_user_article(user_email=request.user, slug=slug)
+        serializer = self.serializer_class(article, data=request.data['article'])
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(
+            {"article": serializer.data,
+             "message": "Article successfully updated"},
+            status=status.HTTP_200_OK
+                        )
```
Comment (MuhanguziDavid) [inline authors/apps/articles/views.py]: Align this one as well

File: authors/apps/articles/views.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    def delete(self, request, slug):
+        """Delete a single article
+        :params request slug
+        :return article"""
+        Article.delete_article(user_email=request.user, slug=slug)
+
+        return Response(
+            {"message": "Article has been deleted"},
+            status=status.HTTP_204_NO_CONTENT
+                        )
```
Comment (MuhanguziDavid) [inline authors/apps/articles/views.py]: Align this one as well

**Review by RonKbS — APPROVED** [review]
This should be showing that no article was found for when trying to update a non-existent article but otherwise LGTM.

screen shot 2018-10-10 at 20 24 17

**Comment by mendozabree** [comment]
This is the intended behaviour of the endpoint.
